File: bazel_ros2_rules/lib/ros_environment/unique.cc
#include <stdlib.h>
#include <unistd.h>
#include <algorithm>

#include "unique.h"

namespace bazel_ros2_rules {

namespace {

std::string replace(std::string str, const std::string& from, const std::string& to) {
  size_t pos = 0;
  while ((pos = str.find(from, pos)) != std::string::npos) {
    str.replace(pos, from.length(), to);
    pos += to.length();
  }
  return str;
}

std::string ltrim(std::string str, char pad = ' ') {
  str.erase(str.begin(), std::find_if(str.begin(), str.end(), [&](char c) {
    return c != pad;
  }));
  return str;
}

}  // namespace

void EnforceUniqueROSEnvironment(
    std::optional<std::string> unique_identifier,
    std::optional<std::string> scratch_directory,
    std::optional<std::filesystem::path> temp_dir)
{
  if (!unique_identifier.has_value()) {
    const char* test_target = getenv("TEST_TARGET");
    if (test_target != nullptr) {
      unique_identifier = ltrim(replace(test_target, "//", "/"), '/');
    } else {
      unique_identifier = std::to_string(getpid());
    }
  }

  if (!temp_dir.has_value()) {
    const char* test_tmpdir = getenv("TEST_TMPDIR");
    if (test_tmpdir != nullptr) {
      temp_dir = std::filesystem::path(test_tmpdir);
    } else {
      temp_dir = std::filesystem::temp_directory_path();
    }
  }

  if (!scratch_directory.has_value()) {
    scratch_directory = (temp_dir.value() / unique_identifier.value()).string();
    if (!std::filesystem::is_directory(scratch_directory.value())) {
      std::filesystem::create_directories(scratch_directory.value());
    }
  }

  // ROS wants to write text logs, so point it to the temporary test directory via ROS_HOME.
  const std::filesystem::path ros_home_directory = 
    std::filesystem::path(scratch_directory.value()) / "ros";
  setenv("ROS_HOME", ros_home_directory.c_str(), 1);
}

}  // namespace bazel_ros2_rules

File: bazel_ros2_rules/lib/ros_environment/unique.py
import os


def make_unique_ros_environment(
    *,
    unique_identifier=None,
    scratch_directory=None,
    temp_dir=None,
):
    """
    Generates a unique ROS 2 environment for tests and processes.

    Note this environment does not imply isolation. 
    See //lib/network_isolation for same host isolation.  

    Warning:
        scratch_directory should generally be unique to prevent collisions
        among environments intended to be distinct.
    """  # noqa

    if unique_identifier is None:
        is_bazel_test = "TEST_TARGET" in os.environ
        # A PID should be unique for a single machine if we are not
        # running inside `bazel test`. This may or may not be true
        # inside of `bazel test - it is explicitly undefined.
        # https://bazel.build/reference/test-encyclopedia#initial-conditions
        if is_bazel_test:
            target_name = os.environ["TEST_TARGET"]
            unique_identifier = target_name.replace("//", "/")
            unique_identifier = unique_identifier.lstrip("/")
        else:
            unique_identifier = str(os.getpid())

    if temp_dir is None:
        temp_dir = os.environ.get("TEST_TMPDIR", "/tmp")

    if scratch_directory is None:
        scratch_directory = os.path.join(temp_dir, unique_identifier)
        if not os.path.isdir(scratch_directory):
            os.makedirs(scratch_directory, exist_ok=True)

    assert unique_identifier is not None
    assert scratch_directory is not None
    assert os.path.isdir(scratch_directory), scratch_directory
    env = {}
    # ROS wants to write text logs, so point it to the temporary
    # test directory via ROS_HOME.
    env["ROS_HOME"] = os.path.join(scratch_directory, "ros")
    return env


def enforce_unique_ros_environment(**kwargs):
    os.environ.update(make_unique_ros_environment(**kwargs))
